### ISSUE TITLE
Suggestion: 'onStop()' should be called *after* all sessions 'onEnded()' have been called

### ISSUE THREAD

**Issue by daattali** [rstudio/shiny#3122]
The `onStop()` function registers a callback to be called when the app is stopped/killed/dies. The `session$onSessionEnded()` (or `session$onEnded()` or `onStop(session != NULL)`) function registers a callback to be called when a specific session is stopped.

I haven't tested in other environments, but on my Windows RStudio IDE, when an app dies, the app stop callback is called before the session callbacks. 

For example:

```
library(shiny)
onStop(function() message("app stopped"))
ui <- fluidPage()
server <- function(input, output, session) {
  session$onSessionEnded(function() message("session ended"))
}
shinyApp(ui, server)
```

When you kill the app, you'll see the order of callbacks printed to the console.

I consider the sessions to be "children" of the "parent" app, and IMO the children should have a chance to cleanup before the parent does.

One example where this could become an issue is when you have a database connection that gets closed in `onStop()`. The individual sessions may want to record something when they end, but if the app callback happens before, then they cannot.

**Comment by wch** [comment]
I think this is a duplicate of #3081.

**Comment by daattali** [comment]
Yes you're right, in that issue you mention "two things wrong", and this seems to be the second one. I'll let you decide if to close or keep this open, whether you want to treat each one as a separate GH issue or not
